### PR TITLE
feat: faulty provider detection for tvwap

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

Adds faulty provider detection logic for tvwap. Also includes some cleanup on the StandardDeviation method & a couple type aliases to make code easier to read. 

progress on: #542

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/oracle.go [hunk 1]
```diff
@@ -352,28 +361,37 @@ func (o *Oracle) getOrSetProvider(ctx context.Context, providerName string) (pro
...
+		}
+		for base, price := range providerPrices {
+			priceMap[providerName][base] = price.Price
+		}
+	}
+
+	deviations, means, err := StandardDeviation(priceMap)
 	if err != nil {
-		return make(map[string]map[string]provider.TickerPrice), nil
+		return make(provider.AggregatedProviderPrices), nil
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:386]: why not return the error?

File: price-feeder/oracle/oracle.go [hunk 2]
```diff
@@ -352,28 +361,37 @@ func (o *Oracle) getOrSetProvider(ctx context.Context, providerName string) (pro
...
-// filterDeviations find the standard deviations of the prices of
-// all assets, and filter out any providers that are not within 2𝜎 of the mean.
-func (o *Oracle) filterDeviations(
-	prices map[string]map[string]provider.TickerPrice) (
-	map[string]map[string]provider.TickerPrice, error,
+// filterTickerDeviations finds the standard deviations of the prices of
+// all assets, and filters out any providers that are not within 2𝜎 of the mean.
+func (o *Oracle) filterTickerDeviations(
+	prices provider.AggregatedProviderPrices) (
+	provider.AggregatedProviderPrices, error,
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:368]: If this is too long, it should be in the format of:

```go
func foo(
  arg1,
  arg2,
  ...
) (x, y, z) {

}
```

File: price-feeder/oracle/oracle.go [hunk 3]
```diff
@@ -393,6 +411,67 @@ func (o *Oracle) filterDeviations(
...
+			if _, ok := tvwaps[providerName]; !ok {
+				tvwaps[providerName] = make(map[string]sdk.Dec)
+			}
+			tvwaps[providerName][assetName] = asset
+		}
+	}
+
+	deviations, means, err := StandardDeviation(tvwaps)
+	if err != nil {
+		return nil, err
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:450]: so here we return `nil, err` which is what I'd expect, but for `filterTickerDeviations` we don't seem to do this?

File: price-feeder/oracle/oracle.go [hunk 4]
```diff
@@ -352,28 +361,36 @@ func (o *Oracle) getOrSetProvider(ctx context.Context, providerName string) (pro
...
+
+	deviations, means, err := StandardDeviation(priceMap)
 	if err != nil {
-		return make(map[string]map[string]provider.TickerPrice), nil
+		return nil, err
 	}
 
-	// Accept any prices that are within 2𝜎, or for which we couldn't get 𝜎
+	// accept any prices that are within 2𝜎, or for which we couldn't get 𝜎
 	for providerName, priceMap := range prices {
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:389]: nit: let's not shadow the `priceMap` variable here. It makes it hard to reason about what variable we're using.

Comment (adamewozniak) [reply price-feeder/oracle/oracle.go]: Updated @ 2c424a51a9165a335b09f926305e767b03bf9cc9

File: price-feeder/oracle/oracle.go [hunk 5]
```diff
@@ -393,6 +407,66 @@ func (o *Oracle) filterDeviations(
...
+			if _, ok := deviations[base]; !ok ||
+				(price.GTE(means[base].Sub(deviations[base].Mul(deviationThreshold))) &&
+					price.LTE(means[base].Add(deviations[base].Mul(deviationThreshold)))) {
+				if _, ok := filteredCandles[providerName]; !ok {
+					filteredCandles[providerName] = make(map[string][]provider.CandlePrice)
+				}
+				filteredCandles[providerName][base] = candles[providerName][base]
+			} else {
+				telemetry.IncrCounter(1, "failure", "provider")
+				o.logger.Warn().Str("base", base).Str("provider", providerName).Msg(
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:460]: should we also log the price?

File: price-feeder/oracle/oracle.go [hunk 6]
```diff
@@ -393,6 +407,66 @@ func (o *Oracle) filterDeviations(
...
+		candlePrices := make(provider.AggregatedProviderCandles)
+
+		for assetName, asset := range c {
+			if _, ok := candlePrices[providerName]; !ok {
+				candlePrices[providerName] = make(map[string][]provider.CandlePrice)
+			}
+			candlePrices[providerName][assetName] = asset
+		}
+
+		tvwap, err := ComputeTVWAP(candlePrices)
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:430]: So we call `ComputeTVWAP` when filtering and also after we've filtered?

Comment (adamewozniak) [reply price-feeder/oracle/oracle.go:430]: Yup, we calculate tvwap for each provider, use that to average & filter out, then calculate tvwap for all providers together - it's not that efficient but seemed the most clean, any ideas for how to change it?

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**